ledger/src/blockstore_processor.rs

```diff
@@ -1049,7 +1049,9 @@ fn process_bank_0(
     )
     .expect("processing for bank 0 must succeed");
     bank0.freeze();
-    blockstore.insert_bank_hash(bank0.slot(), bank0.hash(), false);
+    if blockstore.is_primary_access() {
+        blockstore.insert_bank_hash(bank0.slot(), bank0.hash(), false);
```

Comment:
If the change looks good to @carllin, we may also consider adding an assert inside insert_bank_hash().

Reply:
This thought did cross my mind, but I opted to leave the check outside of the blockstore functions since the secondary use case is so limited. My thought process was that if we added a check to any of the modifying functions, we should arguably add a check to all modifying blockstore functions

Reply:
@steviez I see your point, although enforcing this new check on every new change in the future might be hard to maintain.

One alternative is to pass that access mode down to the `LedgerColumn`: https://github.com/solana-labs/solana/blob/7ba66eed0e93a5420acc0d509f8e460163835a31/ledger/src/blockstore_db.rs#L964 and have it no-op when those modes are enabled

Reply:
Another thing of note, right now when we open as a secondary we run this code: https://github.com/solana-labs/solana/blob/7ba66eed0e93a5420acc0d509f8e460163835a31/ledger/src/blockstore_db.rs#L336-L341, which internally in the wrapper calls here: https://docs.rs/rocksdb/latest/src/rocksdb/db.rs.html#448-451, passing `AccessType::Secondary`

What if instead we opened with the readonly option here: https://docs.rs/rocksdb/latest/rocksdb/struct.DBWithThreadMode.html#method.open_cf_for_read_only

Reply:
`AccessType::Secondary` in RocksDB implicitly means that it is read-only.

> Perhaps then it will cause all writes to fail at runtime in which case we know which functions to go and write handlers for the failures for,

Given my earlier note, this already happens with `Secondary`. The new places that I added `is_primary_access()` guards around were found through CI and running several `solana-ledger-tool` commands locally, and noting which spots spit a RocksDB error back at me.

Reply:
> and have it return a failure when those modes are enabled. Annoying bit there is all existing code will have to handle that failure.

I don't think this would provide any extra value - RocksDB seems to already do a similar check under the hood and will spit an error out about the operation not being supported in Secondary mode. The operations on `LedgerColumn` return a Result, so all of our functions either bubble that Result up with `?` operator or do an `unwrap()` in the function

Reply:
Got it, that makes sense. I think we should rework some of the existing `blockstore_processor` (tests with sufficient code coverage of the entire restart logic starting at `process_blockstore_from_root()` to take a db with a `AccessType::Secondary` so that we can catch these types of failures in the future.

Reply:
> Got it, that makes sense. I think we should rework some of the existing `blockstore_processor` (tests with sufficient code coverage of the entire restart logic starting at `process_blockstore_from_root()` to take a db with a `AccessType::Secondary` so that we can catch these types of failures in the future.

That seems reasonable to me